sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/ReadFromKafkaDoFn.java

```diff
@@ -460,7 +460,28 @@ public ProcessContinuation processElement(
           }
           return ProcessContinuation.resume();
         }
+        long skippedRecords = 0L;
+        final Stopwatch sw = Stopwatch.createStarted();
         for (ConsumerRecord<byte[], byte[]> rawRecord : rawRecords) {
+          // If the Kafka consumer returns a record with an offset that is already processed
+          // the record can be safely skipped. This is needed because there is a possibility
+          // that the seek() above fails to move the offset to the desired position. In which
+          // case poll() would return records that are already cnsumed.
+          if (rawRecord.offset() < startOffset) {
+            // If the start offset is not reached even after skipping the records for 10 seconds
+            // then the processing is stopped with a backoff to give the Kakfa server some time
+            // catch up.
+            if (sw.elapsed().getSeconds() > 10L) {
+              LOG.error(
+                  "The expected offset was not reached even after skipping consumed records for 10 seconds,"
+                      + " the processing of this bundle will be attempted at a later time.");
+              return ProcessContinuation.resume()
+                  .withResumeDelay(org.joda.time.Duration.standardSeconds(10L));
+            } else {
+              skippedRecords++;
+              continue;
+            }
+          }
           if (!tracker.tryClaim(rawRecord.offset())) {
```

Comment:
I think you should have 
if (skippedRecords > 0) {
  LOG.warn(....) // about skipping
}
here instead.

Below you are going to log it separately for every batch of rawRecords and you also will log it for skipping 0 records.

Reply:
Moved the log here, but I don't think the if statement is needed here. Since the skippedRecord is always going to be > 0 if it reaches the logging statement.

Reply:
it is now going to log for every skipped record. I meant to move it to the same scope as the tryclaim

```
    ++skippedRecords;
    continue;
  }
}
if (skippedRecords > 0) {
  // log about skipped records since we're done skipping (didn't continue above)
  LOG()...
  skippedRecords = 0;  // so we don't log again
}
if (!tracker.tryClaim(rawRecord.offset())) {
```

Reply:
Even in the case that you suggested, it will print a LOG statement for every record that the poll returned once 1 record is skipped. That's the reason I had the LOG outside the for loop to only print the statement once. Let me know if that makes sense.

Reply:
You can set skippedRecords to 0 to prevent that, see the code snippet above.

Reply:
Done.